### PR TITLE
fire | sandy

### PR DESCRIPTION
**Pull request by crypt0kitty**
# Media Ranker
Congratulations! You're submitting your assignment!
## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I am still working on this
Describe how you approached testing that model method. What edge cases did you come up with? | I used the examples from the: Unit Testing Model to test all/presence/uniqueness validations along with relations. I still need to add edge cases. 
What are `session` and `flash`? What is the difference between them? | Sessions have more to do with the browser state and flash/flash.now lets a user know of an error or success via the controller.
What was one thing that you gained more clarity on through this assignment? | How to use ruby code within HTML. I probably spent too much time trying to get my herkou site to look exactly like the example. I learned the hard way that I am not at a point where I can code with 1) perfect styling 2) perfect syntax and 3) perfect testing in just one week.
What is the Heroku URL of your deployed application? | https://cryptic-beyond-69599.herokuapp.com/

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+      end
+
+      flash[:success] = "Successfully logged in as new user: #{user.name}"
+
+
+      session[:user_id] = user.id #setting the session id to the user_id
+      flash[:success] = "Successfully logged in as returning user: #{user.name}"
+    else
+      #new user
```
Comment (kaidamasaki) [inline app/controllers/users_controller.rb]: This is the returning user case.

**Review by kaidamasaki — COMMENTED** [review]
Your submission was largely incomplete.  Much of what is here looks good but just looks unfinished.  

The main learning goal I'd like to see you hit going forward is on model testing.

# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | Doesn't limit to 10 pieces of media.
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | Broken view.
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | Not sorted by vote.
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message |  No details about error.
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | Hardcoded to a single user.
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | Hardcoded to a single user.
| 13. Verify that number of votes determines the Media Spotlight | 
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | 
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | 
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | 
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | 

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | 
| 3. Practices git with at least 25 small commits and meaningful commit messages | Only 22 commits.

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | 
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | 
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | 
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | No top ten method.
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | 
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | No fixtures.
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | No relationship tests.
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | No relationship tests.
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | No business logic tests.

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | 
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | 
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | ✔️

File: app/models/vote.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
```
Comment (kaidamasaki) [inline app/models/vote.rb:3]: You probably wanted a validation here to prevent users from voting twice.

File: app/controllers/homepages_controller.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class HomepagesController < ApplicationController
+  def index
+    @works = Work.all
+    @movies = Work.all.select { |w| w.category == "movie" }
+    @books = Work.all.select { |w| w.category == "book" }
+    @albums = Work.all.select { |w| w.category == "album" }
```
Comment (kaidamasaki) [inline app/controllers/homepages_controller.rb]: Instead of returning all of the works here you should only return up to 10 of each category.

File: app/views/users/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+      <th>Votes</th>
+      <th>Joined</th>
+    </tr>
+    </thead>
+    <tbody>
+    <tr>
+      <td><a href="/users/615">dee</a></td>
+      <td>0</td>
+      <td>May 13, 2020</td>
+    </tr>
```
Comment (kaidamasaki) [inline app/views/users/index.html.erb:16]: You should have loaded the `User`s instead of hardcoding this here.

File: app/models/work.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+class Work < ApplicationRecord
+  has_many :votes
+  validates :title, presence: true, uniqueness: true
+
+  def self.spotlight
+    spotlight_for_all_works = Work.all.sample
+    return spotlight_for_all_works
+  end
```
Comment (kaidamasaki) [inline app/models/work.rb:8]: Your spotlight should have only sampled from the works with the most votes.

File: app/views/homepages/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,54 @@
+<main>
+  <section class="spotlight">
+    <h2 class="spotlight__header">
+      <span class="spotlight__header--prefix">Media Spotlight</span>
+      <a class="spotlight__link-to" href="works/605.html"><%=@spotlight.description%></a> <br>
```
Comment (kaidamasaki) [inline app/views/homepages/index.html.erb:5]: This should have linked to the work the was chosen to be spotlighted, instead of being hardcoded to work 605.

File: app/models/work.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+class Work < ApplicationRecord
```
Comment (kaidamasaki) [inline app/models/work.rb:1]: You should have had a `self.top_ten` method here.

**Review by kaidamasaki — COMMENTED** [review]
You made really good progress here.  It's still red but it's _so close_ to being yellow.  If you'd gotten one more piece of functionality the assignment would have been yellow.

# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | Not sorted by vote.
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | Only lists hardcoded user.
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | Only lists hardcoded user.
| 13. Verify that number of votes determines the Media Spotlight | Media spotlight is random.
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | Votes table is not updated.
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | Can vote on the same work twice.
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | No top ten method.
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | No filter for finding a work.
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | No fixtures.
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | Tests exist but are failing.
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | No business logic tests.

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | 
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | ✔️
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |
